// @po-ui/ng-components/esm2020/lib/components/po-navbar/po-navbar-logo/po-navbar-logo.component.mjs
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function PoNavbarLogoComponent_img_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "img", 2);
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("src", ctx_r0.logo, i0.ɵɵsanitizeUrl);
} }
export class PoNavbarLogoComponent {
}
PoNavbarLogoComponent.ɵfac = function PoNavbarLogoComponent_Factory(t) { return new (t || PoNavbarLogoComponent)(); };
PoNavbarLogoComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoNavbarLogoComponent, selectors: [["po-navbar-logo"]], inputs: { logo: ["p-logo", "logo"] }, decls: 2, vars: 1, consts: [["href", "./"], ["class", "po-navbar-logo-image", "alt", "logo", 3, "src", 4, "ngIf"], ["alt", "logo", 1, "po-navbar-logo-image", 3, "src"]], template: function PoNavbarLogoComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "a", 0);
        i0.ɵɵtemplate(1, PoNavbarLogoComponent_img_1_Template, 1, 1, "img", 1);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.logo);
    } }, dependencies: [i1.NgIf], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoNavbarLogoComponent, [{
        type: Component,
        args: [{ selector: 'po-navbar-logo', template: "<a href=\"./\">\n  <img *ngIf=\"logo\" class=\"po-navbar-logo-image\" alt=\"logo\" [src]=\"logo\" />\n</a>\n" }]
    }], null, { logo: [{
            type: Input,
            args: ['p-logo']
        }] }); })();
