// @po-ui/ng-components/esm2020/lib/components/po-table/po-table-column-label/po-table-column-label.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./../../../services/po-color-palette/po-color-palette.service";
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para a criação da representação da legenda, em formato de texto .
 */
export class PoTableColumnLabelComponent {
    constructor(poColorPaletteService) {
        this.poColorPaletteService = poColorPaletteService;
    }
    /** Objeto com os dados do label */
    set value(value) {
        if (value) {
            value.color = this.poColorPaletteService.getColor(value);
        }
        this._value = value;
    }
    get value() {
        return this._value;
    }
}
PoTableColumnLabelComponent.ɵfac = function PoTableColumnLabelComponent_Factory(t) { return new (t || PoTableColumnLabelComponent)(i0.ɵɵdirectiveInject(i1.PoColorPaletteService)); };
PoTableColumnLabelComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoTableColumnLabelComponent, selectors: [["po-table-column-label"]], inputs: { value: ["p-value", "value"] }, decls: 2, vars: 3, template: function PoTableColumnLabelComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "span");
        i0.ɵɵtext(1);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵclassMap("po-table-column-label po-" + (ctx.value == null ? null : ctx.value.color));
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1(" ", ctx.value == null ? null : ctx.value.label, "\n");
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoTableColumnLabelComponent, [{
        type: Component,
        args: [{ selector: 'po-table-column-label', changeDetection: ChangeDetectionStrategy.OnPush, template: "<span [class]=\"'po-table-column-label po-' + value?.color\">\n  {{ value?.label }}\n</span>\n" }]
    }], function () { return [{ type: i1.PoColorPaletteService }]; }, { value: [{
            type: Input,
            args: ['p-value']
        }] }); })();
