// @po-ui/ng-templates/esm2020/lib/components/po-page-blocked-user/po-page-blocked-user-contacts/po-page-blocked-user-contacts.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewChild } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = ["contactGroup"];
const _c1 = ["mailItem"];
const _c2 = ["phoneItem"];
const poPageBlockedUserContactItemMargin = 16;
export class PoPageBlockedUserContactsComponent {
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
        this.overflowItem = true;
    }
    set email(value) {
        this._email = value;
        this.checkContactItemWidth();
    }
    get email() {
        return this._email;
    }
    set phone(value) {
        this._phone = value;
        this.checkContactItemWidth();
    }
    get phone() {
        return this._phone;
    }
    checkContactItemWidth() {
        this.overflowItem = true;
        if (this.phone && this.email) {
            this.changeDetector.detectChanges();
            const phoneWidth = this.phoneItem.nativeElement.offsetWidth;
            const mailWidth = this.mailItem.nativeElement.offsetWidth;
            const contactGroupHalfWidth = this.contactGroup.nativeElement.offsetWidth / 2 - poPageBlockedUserContactItemMargin;
            this.overflowItem =
                phoneWidth > contactGroupHalfWidth || mailWidth > contactGroupHalfWidth - poPageBlockedUserContactItemMargin;
        }
    }
}
PoPageBlockedUserContactsComponent.ɵfac = function PoPageBlockedUserContactsComponent_Factory(t) { return new (t || PoPageBlockedUserContactsComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
PoPageBlockedUserContactsComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoPageBlockedUserContactsComponent, selectors: [["po-page-blocked-user-contacts"]], viewQuery: function PoPageBlockedUserContactsComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 7);
        i0.ɵɵviewQuery(_c1, 7);
        i0.ɵɵviewQuery(_c2, 7);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.contactGroup = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.mailItem = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.phoneItem = _t.first);
    } }, inputs: { email: ["p-email", "email"], phone: ["p-phone", "phone"] }, decls: 16, vars: 14, consts: [[1, "po-page-blocked-user-contact-group", "po-row", "po-mb-3"], ["contactGroup", ""], [1, "po-page-blocked-user-contact-item", 3, "ngClass"], ["target", "_self", 1, "po-page-blocked-user-link", "po-clickable", 3, "href"], [1, "po-page-blocked-user-contact-group-item"], [1, "po-page-blocked-user-contact-icon", "po-icon", "po-icon-telephone", "po-pr-1"], [1, "po-page-blocked-user-contact-text", "po-font-text"], ["phoneItem", ""], [1, "po-page-blocked-user-contact-item", "po-page-blocked-user-contact-mail", 3, "ngClass"], [1, "po-page-blocked-user-contact-icon", "po-icon", "po-icon-mail", "po-pr-1"], ["mailItem", ""]], template: function PoPageBlockedUserContactsComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0, 1)(2, "div", 2)(3, "a", 3)(4, "div", 4);
        i0.ɵɵelement(5, "span", 5);
        i0.ɵɵelementStart(6, "div", 6, 7);
        i0.ɵɵtext(8);
        i0.ɵɵelementEnd()()()();
        i0.ɵɵelementStart(9, "div", 8)(10, "a", 3)(11, "div", 4);
        i0.ɵɵelement(12, "span", 9);
        i0.ɵɵelementStart(13, "div", 6, 10);
        i0.ɵɵtext(15);
        i0.ɵɵelementEnd()()()()();
    } if (rf & 2) {
        i0.ɵɵadvance(2);
        i0.ɵɵclassProp("po-invisible", !ctx.phone);
        i0.ɵɵproperty("ngClass", ctx.overflowItem ? "po-md-12" : "po-md-6");
        i0.ɵɵadvance(1);
        i0.ɵɵclassProp("po-page-blocked-user-contact-content-inline", ctx.overflowItem);
        i0.ɵɵpropertyInterpolate1("href", "tel:", ctx.phone, "", i0.ɵɵsanitizeUrl);
        i0.ɵɵadvance(5);
        i0.ɵɵtextInterpolate(ctx.phone);
        i0.ɵɵadvance(1);
        i0.ɵɵclassProp("po-invisible", !ctx.email);
        i0.ɵɵproperty("ngClass", ctx.overflowItem ? "po-md-12" : "po-md-6");
        i0.ɵɵadvance(1);
        i0.ɵɵclassProp("po-page-blocked-user-contact-content-inline", ctx.overflowItem);
        i0.ɵɵpropertyInterpolate1("href", "mailto:", ctx.email, "", i0.ɵɵsanitizeUrl);
        i0.ɵɵadvance(5);
        i0.ɵɵtextInterpolate(ctx.email);
    } }, dependencies: [i1.NgClass], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageBlockedUserContactsComponent, [{
        type: Component,
        args: [{ selector: 'po-page-blocked-user-contacts', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div #contactGroup class=\"po-page-blocked-user-contact-group po-row po-mb-3\">\n  <div\n    class=\"po-page-blocked-user-contact-item\"\n    [class.po-invisible]=\"!phone\"\n    [ngClass]=\"overflowItem ? 'po-md-12' : 'po-md-6'\"\n  >\n    <a\n      class=\"po-page-blocked-user-link po-clickable\"\n      [class.po-page-blocked-user-contact-content-inline]=\"overflowItem\"\n      href=\"tel:{{ phone }}\"\n      target=\"_self\"\n    >\n      <div class=\"po-page-blocked-user-contact-group-item\">\n        <span class=\"po-page-blocked-user-contact-icon po-icon po-icon-telephone po-pr-1\"></span>\n        <div #phoneItem class=\"po-page-blocked-user-contact-text po-font-text\">{{ phone }}</div>\n      </div>\n    </a>\n  </div>\n  <div\n    class=\"po-page-blocked-user-contact-item po-page-blocked-user-contact-mail\"\n    [class.po-invisible]=\"!email\"\n    [ngClass]=\"overflowItem ? 'po-md-12' : 'po-md-6'\"\n  >\n    <a\n      class=\"po-page-blocked-user-link po-clickable\"\n      [class.po-page-blocked-user-contact-content-inline]=\"overflowItem\"\n      href=\"mailto:{{ email }}\"\n      target=\"_self\"\n    >\n      <div class=\"po-page-blocked-user-contact-group-item\">\n        <span class=\"po-page-blocked-user-contact-icon po-icon po-icon-mail po-pr-1\"></span>\n        <div #mailItem class=\"po-page-blocked-user-contact-text po-font-text\">{{ email }}</div>\n      </div>\n    </a>\n  </div>\n</div>\n" }]
    }], function () { return [{ type: i0.ChangeDetectorRef }]; }, { contactGroup: [{
            type: ViewChild,
            args: ['contactGroup', { static: true }]
        }], mailItem: [{
            type: ViewChild,
            args: ['mailItem', { static: true }]
        }], phoneItem: [{
            type: ViewChild,
            args: ['phoneItem', { static: true }]
        }], email: [{
            type: Input,
            args: ['p-email']
        }], phone: [{
            type: Input,
            args: ['p-phone']
        }] }); })();
