// @po-ui/ng-components/esm2020/lib/components/po-field/po-radio/po-radio-base.component.mjs
import { __decorate } from "tslib";
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { InputBoolean } from '../../../decorators';
import { uuid } from './../../../utils/util';
import { PoRadioSize } from './po-radio-size.enum';
import * as i0 from "@angular/core";
/**
 * @docsPrivate
 *
 * @description
 *
 * O componente `po-radio` é um componente interno e deve ser utilizado em conjunto com o componente `po-radio-group`.
 */
export class PoRadioBaseComponent {
    constructor() {
        /** Define o status do *radio* */
        this.radioValue = false;
        /**
         * @optional
         *
         * @description
         *
         * Define o estado do *radio* como desabilitado;
         *
         * @default `false`
         */
        this.disabled = false;
        /**
         * @optional
         *
         * @description
         *
         * Evento disparado quando o valor do *radio* for alterado.
         */
        this.change = new EventEmitter();
        this.id = uuid();
        this._size = PoRadioSize.medium;
    }
    /**
     * @optional
     *
     * @description
     *
     * Define o tamanho do *radio*
     * @default `medium`
     */
    set size(value) {
        this._size = PoRadioSize[value] ? PoRadioSize[value] : PoRadioSize.medium;
    }
    get size() {
        return this._size;
    }
    changeValue() {
        if (this.propagateChange) {
            this.propagateChange(this.radioValue);
        }
        this.change.emit(this.radioValue);
    }
    checkOption(value) {
        if (this.disabled) {
            return;
        }
        this.changeModelValue(true);
        this.changeValue();
    }
    registerOnChange(fn) {
        this.propagateChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    writeValue(value) {
        if (value !== this.radioValue) {
            this.changeModelValue(value);
        }
    }
}
PoRadioBaseComponent.ɵfac = function PoRadioBaseComponent_Factory(t) { return new (t || PoRadioBaseComponent)(); };
PoRadioBaseComponent.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: PoRadioBaseComponent, inputs: { name: "name", label: ["p-label", "label"], radioValue: ["p-checked", "radioValue"], disabled: ["p-disabled", "disabled"], size: ["p-size", "size"] }, outputs: { change: "p-change" } });
__decorate([
    InputBoolean()
], PoRadioBaseComponent.prototype, "disabled", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoRadioBaseComponent, [{
        type: Directive
    }], null, { name: [{
            type: Input,
            args: ['name']
        }], label: [{
            type: Input,
            args: ['p-label']
        }], radioValue: [{
            type: Input,
            args: ['p-checked']
        }], disabled: [{
            type: Input,
            args: ['p-disabled']
        }], change: [{
            type: Output,
            args: ['p-change']
        }], size: [{
            type: Input,
            args: ['p-size']
        }] }); })();
