// @po-ui/ng-components/esm2020/lib/components/po-field/po-radio/po-radio.component.mjs
import { ChangeDetectionStrategy, Component, forwardRef, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { PoKeyCodeEnum } from './../../../enums/po-key-code.enum';
import { PoRadioBaseComponent } from './po-radio-base.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = ["radioLabel"];
const _c1 = function (a0) { return { "po-radio-large": a0 }; };
const _c2 = function (a0) { return { "po-radio-label-disabled": a0 }; };
/**
 * @docsPrivate
 *
 * @docsExtends PoRadioBaseComponent
 */
export class PoRadioComponent extends PoRadioBaseComponent {
    constructor(changeDetector) {
        super();
        this.changeDetector = changeDetector;
    }
    /**
     * Função que atribui foco ao *radio*.
     *
     * Para utilizá-la é necessário capturar a referência do componente no DOM através do `ViewChild`, como por exemplo:
     *
     * ```
     * import { ViewChild } from '@angular/core';
     * import { PoRadioComponent } from '@po-ui/ng-components';
     *
     * ...
     *
     * @ViewChild(PoRadioComponent, { static: true }) radio: PoRadioComponent;
     *
     * focusRadio() {
     * this.radio.focus();
     * }
     * ```
     *
     */
    focus() {
        if (this.radioLabel && !this.disabled) {
            this.radioLabel.nativeElement.focus();
        }
    }
    onBlur() {
        this.onTouched?.();
    }
    onKeyDown(event, value) {
        if (event.which === PoKeyCodeEnum.space || event.keyCode === PoKeyCodeEnum.space) {
            this.checkOption(value);
            event.preventDefault();
        }
    }
    changeModelValue(value) {
        this.radioValue = typeof value === 'boolean' ? value : false;
        this.changeDetector.detectChanges();
    }
}
PoRadioComponent.ɵfac = function PoRadioComponent_Factory(t) { return new (t || PoRadioComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
PoRadioComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoRadioComponent, selectors: [["po-radio"]], viewQuery: function PoRadioComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.radioLabel = _t.first);
    } }, features: [i0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(() => PoRadioComponent),
                multi: true
            }
        ]), i0.ɵɵInheritDefinitionFeature], decls: 5, vars: 15, consts: [[1, "po-radio-container", 3, "click", "keydown"], ["type", "radio", "role", "radio", 1, "po-radio", 3, "ngClass", "id", "tabindex", "disabled"], ["tabindex", "-1", 1, "po-radio-label", 3, "for", "ngClass"], ["radioLabel", ""]], template: function PoRadioComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵlistener("click", function PoRadioComponent_Template_div_click_0_listener() { return ctx.checkOption(ctx.radioValue); })("keydown", function PoRadioComponent_Template_div_keydown_0_listener($event) { return ctx.onKeyDown($event, ctx.radioValue); });
        i0.ɵɵelement(1, "input", 1);
        i0.ɵɵelementStart(2, "label", 2, 3);
        i0.ɵɵtext(4);
        i0.ɵɵelementEnd()();
    } if (rf & 2) {
        i0.ɵɵattribute("checked", ctx.radioValue);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(11, _c1, ctx.size === "large"))("id", ctx.id)("tabindex", ctx.disabled ? -1 : 0)("disabled", ctx.disabled);
        i0.ɵɵattribute("aria-checked", ctx.radioValue)("aria-label", ctx.label)("aria-disabled", ctx.disabled);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("for", ctx.id)("ngClass", i0.ɵɵpureFunction1(13, _c2, ctx.disabled));
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate1(" ", ctx.label, " ");
    } }, dependencies: [i1.NgClass], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoRadioComponent, [{
        type: Component,
        args: [{ selector: 'po-radio', changeDetection: ChangeDetectionStrategy.OnPush, providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PoRadioComponent),
                        multi: true
                    }
                ], template: "<div\n  class=\"po-radio-container\"\n  [attr.checked]=\"radioValue\"\n  (click)=\"checkOption(radioValue)\"\n  (keydown)=\"onKeyDown($event, radioValue)\"\n>\n  <input\n    type=\"radio\"\n    [ngClass]=\"{ 'po-radio-large': size === 'large' }\"\n    [attr.aria-checked]=\"radioValue\"\n    [attr.aria-label]=\"label\"\n    [id]=\"id\"\n    class=\"po-radio\"\n    role=\"radio\"\n    [attr.aria-disabled]=\"disabled\"\n    [tabindex]=\"disabled ? -1 : 0\"\n    [disabled]=\"disabled\"\n  />\n  <label\n    #radioLabel\n    [for]=\"id\"\n    class=\"po-radio-label\"\n    [ngClass]=\"{ 'po-radio-label-disabled': disabled }\"\n    tabindex=\"-1\"\n  >\n    {{ label }}\n  </label>\n</div>\n" }]
    }], function () { return [{ type: i0.ChangeDetectorRef }]; }, { radioLabel: [{
            type: ViewChild,
            args: ['radioLabel', { static: true }]
        }] }); })();
