// @po-ui/ng-components/esm2020/lib/components/po-field/po-multiselect/po-multiselect-item/po-multiselect-item.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente que constrói cada item no dropdown, contendo o checkbox e o label.
 */
export class PoMultiselectItemComponent {
    constructor() {
        /** Esta propriedade indica se o campo está selecionado, indenterminate ou não. */
        this.selected = false;
        /** Evento que será disparado toda vez que o usuário marcar ou desmarcar um item. */
        this.change = new EventEmitter();
    }
    itemClicked() {
        this.selected = !this.selected;
        this.change.emit(this.selected);
    }
}
PoMultiselectItemComponent.ɵfac = function PoMultiselectItemComponent_Factory(t) { return new (t || PoMultiselectItemComponent)(); };
PoMultiselectItemComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoMultiselectItemComponent, selectors: [["po-multiselect-item"]], inputs: { label: ["p-label", "label"], selected: ["p-selected", "selected"] }, outputs: { change: "p-change" }, decls: 5, vars: 8, consts: [[3, "click"], [1, "po-multiselect-item"], ["type", "checkbox", 1, "po-multiselect-checkbox-input", 3, "checked"], [1, "po-multiselect-checkbox-label", "po-clickable"]], template: function PoMultiselectItemComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "li", 0);
        i0.ɵɵlistener("click", function PoMultiselectItemComponent_Template_li_click_0_listener() { return ctx.itemClicked(); });
        i0.ɵɵelementStart(1, "a", 1);
        i0.ɵɵelement(2, "input", 2);
        i0.ɵɵelementStart(3, "label", 3);
        i0.ɵɵtext(4);
        i0.ɵɵelementEnd()()();
    } if (rf & 2) {
        i0.ɵɵclassProp("po-multiselect-item-selected", ctx.selected);
        i0.ɵɵadvance(2);
        i0.ɵɵclassProp("po-multiselect-checkbox-input-indeterminate", ctx.selected === null)("po-multiselect-checkbox-input-checked", ctx.selected);
        i0.ɵɵproperty("checked", ctx.selected);
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate1(" ", ctx.label, " ");
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoMultiselectItemComponent, [{
        type: Component,
        args: [{ selector: 'po-multiselect-item', changeDetection: ChangeDetectionStrategy.OnPush, template: "<li [class.po-multiselect-item-selected]=\"selected\" (click)=\"itemClicked()\">\n  <a class=\"po-multiselect-item\">\n    <input\n      [checked]=\"selected\"\n      class=\"po-multiselect-checkbox-input\"\n      [class.po-multiselect-checkbox-input-indeterminate]=\"selected === null\"\n      [class.po-multiselect-checkbox-input-checked]=\"selected\"\n      type=\"checkbox\"\n    />\n\n    <label class=\"po-multiselect-checkbox-label po-clickable\">\n      {{ label }}\n    </label>\n  </a>\n</li>\n" }]
    }], null, { label: [{
            type: Input,
            args: ['p-label']
        }], selected: [{
            type: Input,
            args: ['p-selected']
        }], change: [{
            type: Output,
            args: ['p-change']
        }] }); })();
